### PR TITLE
Launch smoke tests via ECS

### PR DESCRIPTION
**Pull request by v-stickykeys**
### Summary

Launching smoke tests with ECS instead of lambda because of https://github.com/3box/ceramic-smoke-tests/pull/30

### Motivation

When we upgraded our Ceramic packages to ESM, smoke tests broke in lambda. Dockerizing them and running on ECS makes it easier to handle upgrades. (More details in PR referenced above).

### Changes

- Replaced API gateway invocation with launching an ECS task.
- Renamed `TaskIdParam` const to be prefixed with the job name

### CODE REVIEW

File: cd/manager/jobs/anchor.go [hunk 1]
```diff
@@ -9,8 +9,9 @@ import (
 	"github.com/3box/pipeline-tools/cd/manager"
 )
 
+// Allow up to 3 hours for anchor workers to run
 const AnchorStalledTime = 3 * time.Hour
-const TaskIdParam = "id"
+const AnchorTaskIdParam = "id"
```
Comment (smrz2001) [inline cd/manager/jobs/anchor.go]: Mind adding this as a literal under this section in `models.go`?
```
const (
	JobParam_Component string = "component"
	JobParam_Sha       string = "sha"
	JobParam_Error     string = "error"
	JobParam_Layout    string = "layout"
	JobParam_Manual    string = "manual"
	JobParam_Force     string = "force"
	JobParam_Start     string = "start"
	JobParam_Rollback  string = "rollback"
	JobParam_Delayed   string = "delayed"
	JobParam_Stalled   string = "stalled"
	JobParam_Source    string = "source"
)
```
`id` is generic enough to be useful in other places, so maybe `JobParam_Id        string = "id"`.

File: cd/manager/jobs/smoke.go [hunk 2]
```diff
@@ -9,33 +9,72 @@ import (
 	"github.com/3box/pipeline-tools/cd/manager"
 )
 
+// Allow up to 1 hours for smoke tests to run
+const SmokeTestFailureTime = 1 * time.Hour
+const SmokeTaskIdParam = "id"
```
Comment (smrz2001) [inline cd/manager/jobs/smoke.go]: Can reuse the new `id` literal here as well.

File: cd/manager/jobs/smoke.go [hunk 3]
```diff
@@ -9,33 +9,72 @@ import (
 	"github.com/3box/pipeline-tools/cd/manager"
 )
 
+// Allow up to 1 hours for smoke tests to run
+const SmokeTestFailureTime = 1 * time.Hour
```
Comment (smrz2001) [inline cd/manager/jobs/smoke.go]: Is this too long? Lambda test timeout was 5 minutes, so maybe 10-15 min max here? WDYT?

File: cd/manager/jobs/smoke.go [hunk 4]
```diff
@@ -9,33 +9,72 @@ import (
...
+	return &smokeTestJob{jobState, db, d, notifs, os.Getenv("ENV")}
 }
 
 func (s smokeTestJob) AdvanceJob() (manager.JobState, error) {
 	if s.state.Stage == manager.JobStage_Queued {
-		resourceId := os.Getenv("SMOKE_TEST_RESOURCE_ID")
-		restApiId := os.Getenv("SMOKE_TEST_REST_API_ID")
-		if _, err := s.api.Invoke("GET", resourceId, restApiId, ""); err != nil {
+		// Launch smoke test
+		if id, err := s.d.LaunchTask(
```
Comment (smrz2001) [inline cd/manager/jobs/smoke.go:37]: From the other PR, looks like we won't have an ECS Service at all, and just ECS Tasks for both e2e and smoke tests. Is that correct?

If so, then `e2e.go` also needs to be updated to use `LaunchTask` instead of `LaunchServiceTask`, though that can be a separate PR.

If not, then this call can also be `LaunchServiceTask`, like for e2e tests.

File: cd/manager/jobs/smoke.go [hunk 5]
```diff
@@ -9,33 +9,72 @@ import (
 	"github.com/3box/pipeline-tools/cd/manager"
 )
 
+// Allow up to 1 hours for smoke tests to run
+const SmokeTestFailureTime = 1 * time.Hour
+const SmokeTaskIdParam = "id"
+
+const ECSCluster = "ceramic-qa-tests"
```
Comment (smrz2001) [inline cd/manager/jobs/smoke.go]: Might be worth moving these to `models.go` too? Just a thought.
```
const TestsCluster = "ceramic-qa-tests"
const TestsNetworkConfigurationParameter = "network_configuration"

const (
    TestType_Smoke string = "smoke"
    TestType_E2e string = "e2e"
)
```
and then the code in `smoke.go` becomes:
```
if id, err := s.d.LaunchTask(
	manager.TestsCluster,
	manager.TestsCluster+"-"+manager.TestType_Smoke+"-"+s.env,
	manager.TestsCluster+"-"+manager.TestType_Smoke,
	"/"+manager.TestsCluster+"-"+manager.TestType_Smoke+"/"+manager.TestsNetworkConfigurationParameter,
	nil); err != nil {
```
Something like that. Just a thought. Could also use `fmt.sprintf` to make the string formats a little more explicit and readable. e.g.
```
fmt.Sprintf("%s-%s-%s", manager.TestsCluster, manager.TestType_Smoke, s.env) // for family

fmt.Sprintf("/%s-%s/%s", manager.TestsCluster, manager.TestType_Smoke, manager.TestsNetworkConfigurationParameter) // for n/w config
```

File: cd/manager/jobs/anchor.go (hunk 1 above)
Comment (v-stickykeys) [reply cd/manager/jobs/anchor.go]: will do

File: cd/manager/jobs/smoke.go (hunk 3 above)
Comment (v-stickykeys) [reply cd/manager/jobs/smoke.go]: will set to 15 min

File: cd/manager/jobs/smoke.go (hunk 4 above)
Comment (v-stickykeys) [reply cd/manager/jobs/smoke.go:37]: yes will do it in separate PR because this debt ticket needs to be done first https://linear.app/3boxlabs/issue/PLAT-1052/move-e2e-tests-to-app-tests-in-terraform

File: cd/manager/jobs/smoke.go (hunk 5 above)
Comment (v-stickykeys) [reply cd/manager/jobs/smoke.go]: I think the benefit of having it typed out when you read the code is pretty large

As in, when I see `fmt.Sprintf("%s-%s--%s", manager.TestsCluster, manager.TestType_Smoke, s.env)` it will take me a minute, and flipping back and forth between files, to figure out what the resulting string actually is

vs if I can see `const ECSFamilyPrefix = "ceramic-qa-tests-smoke--"` and `ECSFamilyPrefix+s.env,` on the same page. It's much more obvious what the vaule is.

Comment (v-stickykeys) [reply cd/manager/jobs/smoke.go]: pulling out `const TestsCluster = "ceramic-qa-tests"` makes sense for next pr when we migrate the e2e tests

Comment (smrz2001) [reply cd/manager/jobs/smoke.go]: Ok, sg. Should take `ECS` out of the literal names though to keep the abstractions cleaner. The jobs don't care that ECS is being used under the covers.

**Review by smrz2001 — APPROVED** [review]
LGTM apart from minor editorial comment 🚀
